share/getter.go

```diff
@@ -0,0 +1,98 @@
+package share
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/minio/sha256-simd"
+
+	"github.com/celestiaorg/celestia-node/share/ipld"
+
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+// Getter interface provides a set of accessors for shares by the Root.
+// Automatically verifies integrity of shares(exceptions possible depending on the implementation)
+type Getter interface {
+	// GetShare gets a Share by coordinates in EDS.
+	GetShare(ctx context.Context, root *Root, row, col int) (Share, error)
+
+	// GetEDS gets the full EDS identified by the given root.
+	GetEDS(context.Context, *Root) (*rsmt2d.ExtendedDataSquare, error)
+
+	// GetSharesByNamespace gets all shares from an EDS within the given namespace.
+	// Shares are returned in a row-by-row order if the namespace spans multiple rows.
+	GetSharesByNamespace(context.Context, *Root, namespace.ID) (NamespaceShares, error)
+}
+
+// NamespaceShares represents all the shares with proofs of a specific namespace of an EDS.
+type NamespaceShares []RowNamespaceShares
+
+// RowNamespaceShares represents all the shares with proofs of a specific namespace within a Row of
+// an EDS.
+type RowNamespaceShares struct {
+	Shares []Share
+	Proof  *ipld.Proof
+}
+
+// Flatten returns the concatenated slice of all RowNamespaceShares shares.
+func (ns NamespaceShares) Flatten() []Share {
+	shares := make([]Share, 0)
+	for _, row := range ns {
+		shares = append(shares, row.Shares...)
+	}
+	return shares
+}
+
+// Verify validates NamespaceShares by checking every row with nmt inclusion proof.
+func (ns NamespaceShares) Verify(root *Root, nID namespace.ID) error {
+	originalRoots := make([][]byte, 0)
+	for _, row := range root.RowsRoots {
+		if !nID.Less(nmt.MinNamespace(row, nID.Size())) && nID.LessOrEqual(nmt.MaxNamespace(row, nID.Size())) {
+			originalRoots = append(originalRoots, row)
+		}
+	}
+
+	if len(originalRoots) != len(ns) {
+		return fmt.Errorf("amount of rows differs between root and namespace shares: %v vs %v",
+			len(originalRoots), len(ns))
+	}
+
+	for i, row := range ns {
+		// verify row data against row hash from original root
+		if !row.verify(originalRoots[i], nID) {
+			return fmt.Errorf("row verification failed: %vth row doesn't match original root: %s", i, root.Hash())
```

Comment:
```suggestion
			return fmt.Errorf("row verification failed: row %d doesn't match original root: %s", i, root.Hash())
```

nit but it's not `3th`